### PR TITLE
Feature/search screen

### PR DESCRIPTION
**Pull request by anaselmadhi**
### What has been changed?
- Added a new SearchRecipeScreen composable that serves as a dedicated search screen for recipes, including the UI components for searching and displaying filtered recipes.
- Modified the SearchBar composable to include a search input field with an enhanced UI, aligning with the app’s design for a consistent and modern look.
- Updated the SearchBar functionality to handle text input and filtering directly within the search screen, allowing for a more streamlined and intuitive user experience when searching for recipes.
- Added placeholder elements and basic filtering logic in SearchBar to display search results in real-time.
- Implemented a temporary filtering function within SearchBar to dynamically filter recipes based on the search input.
### Why was this change made?
- This update was made to provide users with a functional and visually appealing search experience within the app, allowing them to find recipes more efficiently.
- Enhancing the search functionality to filter recipes dynamically aligns with the app’s user-centered design principles, making it easier to navigate and locate specific content.
- Updating the SearchBar UI and adding the search screen also improves modularity and reusability, allowing these components to integrate seamlessly into future screens as needed.

### SearchRecipeScreen:
The SearchRecipeScreen composable is designed to serve as the main interface for searching and viewing recipes. It includes a top bar with a centered title and a bottom navigation menu for easy access to other app sections.
The screen dynamically displays the recipe list based on the current search input in the SearchBar.
### SearchBar UI and Functionality:
Added new visual enhancements to SearchBar, including shadow and border styles for better aesthetic integration.
Integrated a filter function in SearchBar to filter recipes based on the user's input. The search term is processed in real-time to display matching recipes in the list below.

### Expanded Testing:
I plan to expand the test coverage to include more comprehensive checks for the search and filter functionality, ensuring accurate filtering and reliable UI interactions.
Additional tests will focus on edge cases for the search input, such as empty inputs, special characters, and case insensitivity.
I will also add tests to validate the behavior of the filter feature once it is fully implemented, ensuring a seamless and intuitive user experience.
### Filter Feature Integration:
In future updates, I will enhance the SearchRecipeScreen by adding a robust Filter feature, allowing users to narrow down recipes by various criteria (e.g., categories, ingredients, dietary restrictions).
This filtering capability will be integrated into the UI with a flexible and expandable filter interface, accessible from the SearchRecipeScreen.
The filter logic will be designed to work in conjunction with the search bar, providing users with a powerful and efficient way to find recipes that meet their specific needs.

Checklist
- [x] Tests added:
Basic UI tests were conducted to confirm that elements are displayed as expected. Currently, no unit tests are implemented for filtering logic, as it is intended as a prototype for UI demonstration only.
- [x] Documentation updated:
Updated inline comments for the SearchRecipeScreen and SearchBar composables to clarify the functionality and outline future improvements.

### CODE REVIEW

**Review by CapraSnitzel — CHANGES_REQUESTED** [review]
First, there a couple of issues found by SonarCloud. You should check them out and correct them.
Then, there are a couple of things going on:
1. You don't have the required coverage in order to merge into main (you have 65% and the required coverage is at least 80%)
2. Another problem is that you did not merge the main branch into your branch. Thus, you don't have the latest files and have conflicts that need resolution.
3. Then there is a problem when going from a small phone to a medium phone:
    - Small phone
    
       SmallPhone
    - Medium phone
      
       MediumPhone
    As you can see, the search bar does not adapt to the format (the filter button disappears). you should use the modifier weight more often. 
In addition, the filter button is very small. This should be fixed as well before merging.
4. The last thing is that, in my opinion, you should put some space between the elements because at the moment, it is very cramped.

So, in order to merge your modifications on the main branch, you should look at the comments as well as the above problems and make the required modifications.

Side note, you should not have multiple commits with the same message and the messages should be more explicit about what was added.

File: app/src/androidTest/java/com/android/sample/e2e/End2EndTest.kt [hunk 1]
```diff
@@ -85,23 +89,28 @@ class EndToEndTest {
...
-      FakeNavHost(navController, userViewModel)
+      FakeNavHost(navController, userViewModel, recipesRepository)
     }
 
     composeTestRule.onNodeWithTag("tabFridge").assertExists().performClick()
     composeTestRule.onNodeWithText("Fridge Screen").assertExists()
 
     composeTestRule.onNodeWithTag("tabSearch").assertExists().performClick()
-    composeTestRule.onNodeWithText("Search Recipe Screen").assertExists()
+    // composeTestRule.onNodeWithText("Search Recipe Screen").assertExists()
```
[the remark this answers is not in the document]
Comment (anaselmadhi) [reply app/src/androidTest/java/com/android/sample/e2e/End2EndTest.kt:99]: Yes, it was written by a developer, but I no longer needed it, so I commented out the line just in case. But we can delete it, indeed

File: app/src/androidTest/java/com/android/sample/e2e/End2EndTest.kt [hunk 2]
```diff
@@ -85,23 +89,28 @@ class EndToEndTest {
...
 
     composeTestRule.onNodeWithTag("tabAddRecipe").assertExists().performClick()
-    composeTestRule.onNodeWithText("Create Recipe Screen").assertExists()
+    // composeTestRule.onNodeWithText("Create Recipe Screen").assertExists()
```
[the remark this answers is not in the document]
Comment (anaselmadhi) [reply app/src/androidTest/java/com/android/sample/e2e/End2EndTest.kt:102]: Same reply (see above)

File: app/src/main/java/com/android/sample/ui/utils/PlateSwipeScaffold.kt
[the remark this answers is not in the document]
Comment (anaselmadhi) [reply app/src/main/java/com/android/sample/ui/utils/PlateSwipeScaffold.kt:1]: see the reply in the next comment.

File: app/src/main/java/com/android/sample/ui/recipe/SearchRecipeScreen.kt
[the remark this answers is not in the document]
Comment (anaselmadhi) [reply app/src/main/java/com/android/sample/ui/recipe/SearchRecipeScreen.kt:1]: Yes, I know, but the scaffold doesn't exactly match the Figma design, so I wanted to make it look closer to the Figma version. I didn't change everything I intended, but it might be a good idea to review this composable to align it perfectly with the Figma design.

File: app/src/main/java/com/android/sample/ui/recipe/SearchRecipeScreen.kt [hunk 3]
```diff
@@ -1,10 +1,171 @@
...
+import com.android.sample.model.recipe.Recipe
+import com.android.sample.model.recipe.RecipesViewModel
+import com.android.sample.resources.C.Tag.SEARCH_BAR_CORNER_RADIUS
+import com.android.sample.ui.navigation.BottomNavigationMenu
+import com.android.sample.ui.navigation.LIST_TOP_LEVEL_DESTINATIONS
 import com.android.sample.ui.navigation.NavigationActions
+import com.android.sample.ui.utils.*
 
+// TODO: quand je cherche et enleve le searchText en le mettant a "" j'ai la liste recipes de base
+// qui est regenerée entière.
```
[the remark this answers is not in the document]
Comment (anaselmadhi) [reply app/src/main/java/com/android/sample/ui/recipe/SearchRecipeScreen.kt:35]: Yes sorry, I forgot to delete it. It was just for me.

File: app/src/main/java/com/android/sample/ui/utils/SearchBar.kt
[the remark this answers is not in the document]
Comment (anaselmadhi) [reply app/src/main/java/com/android/sample/ui/utils/SearchBar.kt:1]: Everything has been implemented according to the Figma design. The search bar design differs across the Figma screens, so I modified the searchBar composable to better match the design and make it more versatile for reuse in multiple composable functions.

File: app/src/main/java/com/android/sample/model/recipe/RecipesViewModel.kt [hunk 4]
```diff
@@ -18,7 +18,7 @@ import okhttp3.OkHttpClient
  *
  * @property repository The repository used to fetch recipe data.
  */
-class RecipesViewModel(private val repository: RecipesRepository) : ViewModel() {
+open class RecipesViewModel(private val repository: RecipesRepository) : ViewModel() {
```
[the remark this answers is not in the document]
Comment (anaselmadhi) [reply app/src/main/java/com/android/sample/model/recipe/RecipesViewModel.kt:21]: Yes it was to create a mock. But I didn't need it. So as you are saying it is not necessary.

File: app/src/main/java/com/android/sample/ui/utils/SearchBar.kt [hunk 5]
```diff
@@ -31,42 +38,53 @@ import com.android.sample.resources.C.Tag.SEARCH_BAR_CORNER_RADIUS
...
+/**
+ * Filters the list of recipes based on the search string.
+ *
+ * @param string the search string to filter the list of recipes.
+ * @param list the list of recipes to filter.
+ * @return the filtered list of recipes.
+ */
+fun filter(string: String, list: List<Recipe>): List<Recipe> {
+  return list.filter { it.strMeal.contains(string) }
 }
```
[the remark this answers is not in the document]
Comment (anaselmadhi) [reply app/src/main/java/com/android/sample/ui/utils/SearchBar.kt:90]: For now, yes, it should be private. However, the function might be useful elsewhere in the future.

File: app/src/main/java/com/android/sample/ui/recipe/SearchRecipeScreen.kt [hunk 6]
```diff
@@ -1,10 +1,171 @@
...
+@Composable
+fun ClearAllChip() {
+  Button(
+      onClick = { /* Clear all filters logic */},
+      colors =
+          ButtonDefaults.buttonColors(
+              containerColor = MaterialTheme.colorScheme.surfaceVariant,
+              contentColor = MaterialTheme.colorScheme.onSurface)) {
+        Text("Clear All", fontSize = 12.sp)
+      }
```
[the remark this answers is not in the document]
Comment (CapraSnitzel) [reply app/src/main/java/com/android/sample/ui/recipe/SearchRecipeScreen.kt:170]: Do you only use them in this class and don't want other to use them? If so, you should make them private

File: app/src/main/java/com/android/sample/ui/recipe/SearchRecipeScreen.kt
[the remark this answers is not in the document]
Comment (CapraSnitzel) [reply app/src/main/java/com/android/sample/ui/recipe/SearchRecipeScreen.kt:1]: I agree with you that on the figma is a different topBar. But as you might have seen, all the other screen have the same topbar (the one in the scaffold). Thus, the fact that the topbar only on one of the search screens might be a mistake and was forgotten to be updated, and maybe it should be good to use the scaffold.
And you have ideas to make the topbar better, share them with us.

File: app/src/main/java/com/android/sample/ui/utils/SearchBar.kt
[the remark this answers is not in the document]
Comment (CapraSnitzel) [reply app/src/main/java/com/android/sample/ui/utils/SearchBar.kt:1]: By looking at the Figma, I don't really see any resemblance between the Figma and you searchBar. The old search bar that was implemented looked much more closer to the design. Maybe it would a good idea to check it out once more.

File: app/src/androidTest/java/com/android/sample/e2e/End2EndTest.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (CapraSnitzel) [reply app/src/androidTest/java/com/android/sample/e2e/End2EndTest.kt:99]: Could you explain why you no longer need it ?

File: app/src/main/java/com/android/sample/model/recipe/RecipesViewModel.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (CapraSnitzel) [reply app/src/main/java/com/android/sample/model/recipe/RecipesViewModel.kt:21]: Then it can be removed :)

File: app/src/androidTest/java/com/android/sample/e2e/End2EndTest.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (anaselmadhi) [reply app/src/androidTest/java/com/android/sample/e2e/End2EndTest.kt:99]: because there si no node with search recipe screen

File: app/src/main/java/com/android/sample/ui/utils/SearchBar.kt [hunk 7]
```diff
@@ -31,7 +32,11 @@ import com.android.sample.resources.C.Tag.SEARCH_BAR_CORNER_RADIUS
  */
 @Preview
 @Composable
-fun SearchBar(modifier: Modifier = Modifier, onValueChange: (String) -> Unit = {}) {
+fun SearchBar(
+    modifier: Modifier = Modifier,
+    onValueChange: (String) -> Unit = {},
+    isSingleLine: Boolean = false
```
Comment (CapraSnitzel) [inline app/src/main/java/com/android/sample/ui/utils/SearchBar.kt:38]: Is the parameter isSingleLine necessary ? Can't you put it directly to true in the composable TextField ? 

In addition, why is it by default to false ? For a search bar, we don't want to write multiple lines, don't we?

File: app/src/main/java/com/android/sample/ui/utils/RecipeList.kt
Comment (CapraSnitzel) [inline app/src/main/java/com/android/sample/ui/utils/RecipeList.kt:1]: Why did you add `${recipe.idMeal}` to the test tags ? There is no reason to update the test tags.

File: app/src/androidTest/java/com/android/sample/e2e/End2EndTest.kt [hunk 8]
```diff
@@ -130,70 +128,50 @@ class EndToEndTest {
...
-
-    composeTestRule.onNodeWithTag("tabAdd Recipe").assertExists().performClick()
-    composeTestRule.onNodeWithText("Create your recipe").assertExists()
-    composeTestRule.onNodeWithTag("tabAdd Recipe").assertExists().performClick()
-    composeTestRule.onNodeWithTag("RecipeTitle").assertExists()
-
-    composeTestRule.onNodeWithTag("tabFridge").assertExists().performClick()
-    composeTestRule.onNodeWithText("Fridge Screen").assertExists()
-  }
-}
```
Comment (CapraSnitzel) [inline app/src/androidTest/java/com/android/sample/e2e/End2EndTest.kt:155]: Is there a reason why this tests was removed ?

File: app/src/androidTest/java/com/android/sample/e2e/End2EndTest.kt [hunk 9]
```diff
@@ -130,70 +128,50 @@ class EndToEndTest {
...
-  val navigationActions = NavigationActions(navController)
-  NavHost(navController = navController, startDestination = Route.SWIPE) {
-    navigation(
-        startDestination = Screen.SWIPE,
-        route = Route.SWIPE,
-    ) {
-      composable(Screen.SWIPE) {
-        SwipePage(navigationActions = navigationActions, recipesViewModel = recipesViewModel)
+  @Composable
+  fun FakeNavHost(
```
Comment (CapraSnitzel) [inline app/src/androidTest/java/com/android/sample/e2e/End2EndTest.kt:135]: Where is this function used ? Because, for the moment, it is not used.

**Review by CapraSnitzel — CHANGES_REQUESTED** [review]
Nice work for correcting the issues from the last review. However, there are still some issues that should be addressed:
1. You have not reached at least 80% coverage. You should try to reach it
2. SonarCloud has brought up some issues. Look at them and try to fix them.
3. I have brought up some problems in the code. Have a look at them and try to fix them.
